src/functional/ffront/ast_passes/fix_missing_locations.py

```diff
@@ -29,23 +30,30 @@ class FixMissingLocations(ast.NodeTransformer):
     some ast nodes and is hence not a replacement for this pass.
     """
 
-    _parent_node: ast.AST
+    _parent_node: list[ast.AST]
 
     @classmethod
     def apply(cls, node: ast.AST) -> ast.AST:
-        return cls(_parent_node=node).visit(node)
+        return cls(_parent_node=[]).visit(node)
 
     def generic_visit(self, node: ast.AST) -> ast.AST:
-        if getattr(node, "lineno", None):
-            self._parent_node = node
+        has_location = hasattr(node, "lineno")
+        if has_location:
+            self._parent_node.append(node)
         else:
-            node.lineno = self._parent_node.lineno
-            node.col_offset = self._parent_node.col_offset
+            node = copy.deepcopy(node)
```

Comment:
`deepcopy` will preserve duplicates, so there's no advantage using `deepcopy` over shallow `copy.copy` (`deepcopy` will also change worst case runtime from O(n) to O(n^2)).

I also think it would be better to have a separate `NodeDeduplication` pass that solves the problem in a general manner. This pass would run before `FixMissingLocations` and establish the pre-condition that there are no duplicate (by identity) nodes in the tree.

Reply:
I switched to a shallow copy.

I see the point of a `NodeDeduplication`, but since this is the only pass that requires it and since it is good practice to copy a node if it is modified based on a property of a parent node I would like to leave it as is.